### ISSUE TITLE
integrate node process with sc stage to request/manufacture items

### ISSUE THREAD

**Issue by raokrutarth** [raokrutarth/manufacturing-os#65]
Integrate the `SuppyChainStage` class with the `SocketBasedNodeProcess` class so items can be requested by the present stage from the upstream nodes and items can be made available to downstream nodes.

**Comment by raokrutarth** [comment]
**Issue**

Since the node objects are fixed after the cluster starts in references like `raft_helper.nodes` or `cluster_flow.nodes`, we cannot add new nodes to the cluster without restarting tho whole cluster. So, nodes cannot find the node metadata (node address, dependency list, etc.) from a `node_id` in the `ClusterWideFlow` API as of today.

**Resolution**

The `ClusterWideFlow` cannot obtain the list of nodes from those determined during `main()` execution (e.g. from the `raft_helper` object's refrences) but has to dynamically keep track of all node objects (not just `node_ids`) as new nodes join/exit the system. 

That is, when a stage asks the `ClusterWideFlow` "To which node address should I send the request for 5  doors of type X" the `ClusterWideFlow` object will return the right node address even if that means the address of a node that just joined the cluster (i.e. it was not part of the cluster object during the cluster bootstrap. That is, the node was added after all other nodes had already started and finished some transactions/transfers between them.)


**Justification for this Functionality**

The SC stage `A` needs the node address to send the `RequestMaterialBatch` message when requesting new raw materials. This address is picked from the `ClusterWideFlow` object's `inbound_edges` field. The `inbound_edges` contain the `node_id`s only as of today and this means, a node that was added after the cluster started will not have it's metadata in node `A`'s in-memory cluster object.

**Comment by raokrutarth** [comment]
The task requires at least 2 people. there are multiple workflows that need to be implemented.

Done so far:

- Full integration with message handler & sc_stage.
- Requests and response definitions.

TODOs of remaining items added.

blocked by a lack of non-async way to get flow.

**Comment by raokrutarth** [comment]
@AndrejSafundzic @chuanqichen @nishantrai18 we still need to merge this logic. This is what "manufactures". What do we need to finish before this can go in?

**Comment by nishantrai18** [comment]
Basically once the leader election and flow consensus are done, this will be unblocked as it'll probably use similar constructs to persist information.
